server/session/handler_enchanting.go

```diff
@@ -0,0 +1,341 @@
+	enchantable, ok := stack.Item().(item.Enchantable)
+	if !ok {
+		// We can't enchant this item.
+		return nil, nil
+	}
+	if len(stack.Enchantments()) > 0 {
+		// We can't enchant this item.
+		return nil, nil
+	}
+
+	// Search for bookshelves around the enchanting table. Bookshelves help boost the value of the enchantments that
+	// are selected, resulting in enchantments that are rarer but also more expensive.
+	random := rand.New(rand.NewSource(s.c.EnchantmentSeed()))
+	bookshelves := searchBookshelves(w, pos)
+	value := enchantable.EnchantmentValue()
+
+	// Calculate the base cost, used to calculate the upper, middle, and lower level costs.
+	baseCost := random.Intn(8) + 1 + (bookshelves >> 1) + random.Intn(bookshelves+1)
+
+	// Calculate the upper, middle, and lower level costs.
+	upperLevelCost := max(baseCost/3, 1)
+	middleLevelCost := baseCost*2/3 + 1
+	lowerLevelCost := max(baseCost, bookshelves*2)
+
+	// Create a list of available enchantments for each slot.
+	return []int{
+			upperLevelCost,
+			middleLevelCost,
+			lowerLevelCost,
+		}, [][]item.Enchantment{
+			createEnchantments(random, stack, value, upperLevelCost),
+			createEnchantments(random, stack, value, middleLevelCost),
+			createEnchantments(random, stack, value, lowerLevelCost),
+		}
+}
+
+// treasureEnchantment represents an enchantment that may be a treasure enchantment.
+type treasureEnchantment interface {
+	item.EnchantmentType
+	Treasure() bool
+}
+
+// createEnchantments creates a list of enchantments for the given item stack and returns them.
+func createEnchantments(random *rand.Rand, stack item.Stack, value, level int) []item.Enchantment {
+	// Calculate the "random bonus" for this level. This factor is used in calculating the enchantment cost, used
+	// during the selection of enchantments.
+	randomBonus := (random.Float64() + random.Float64() - 1.0) * 0.15
+
+	// Calculate the enchantment cost and clamp it to ensure it is always at least one with triangular distribution.
+	cost := level + 1 + random.Intn(value/4+1) + random.Intn(value/4+1)
+	cost = clamp(int(math.Round(float64(cost)+float64(cost)*randomBonus)), 1, math.MaxInt32)
+
+	// Books are applicable to all enchantments, so make sure we have a flag for them here.
+	it := stack.Item()
+	_, book := it.(item.Book)
+
+	// Now that we have our enchantment cost, we need to select the available enchantments. First, we iterate through
+	// each possible enchantment.
+	availableEnchants := make([]item.Enchantment, 0, len(item.Enchantments()))
+	for _, enchant := range item.Enchantments() {
+		if t, ok := enchant.(treasureEnchantment); ok && t.Treasure() {
+			// We then have to ensure that the enchantment is not a treasure enchantment, as those cannot be selected through
+			// the enchanting table.
+			continue
+		}
+
+		// Also make sure that the enchantment is compatible with the item.
+		if book || enchant.CompatibleWithItem(it) {
```

Comment:
Would invert this check to `!book && !compatible` and continue, so that you can move the code afterwards up a level.

Reply:
Inverted.